conbench/entities/history.py

```diff
@@ -570,3 +588,77 @@ def _calculate_z_score(
         z_score = z_score * -1
 
     return z_score
+
+
+def _detect_shifts_with_trimmed_estimators(
+    df: pd.DataFrame, z_score_threshold=5.0
+) -> pd.DataFrame:
+    """Detect outliers and distribution shifts in historical data
+
+    Uses a z-score-based detection algorithm that uses trimmed rolling means and standard
+    deviations to avoid influence by outliers. Takes a dataframe of the same sort as
+    `_add_rolling_stats_columns_to_df`, and returning that dataframe with two additional
+    columns:
+
+    - `is_step` (bool): Is this point the start of a new segment?
+    - `is_outlier` (bool): Is this point an outlier that should be ignored?
+    """
+    tmp_df = copy.deepcopy(df)
+
+    # skip computation if no history
+    if df.shape[0] == 0:
+        tmp_df["is_step"] = pd.Series([], dtype=bool)
+        tmp_df["is_outlier"] = pd.Series([], dtype=bool)
+        return tmp_df
+
+    # pandas likes the data to be sorted
+    tmp_df.sort_values(
+        [
+            "case_id",
+            "context_id",
+            "hash",
+            "repository",
+            "timestamp",
+            "result_timestamp",
```

Comment:
I see, here "result_timestamp" pops up again. Maybe we should add a comment what "timestamp" is :P